vms/evm/sync/customrawdb/sync_progress.go

```diff
@@ -0,0 +1,235 @@
+}
+
+// ClearSyncStorageTrie removes all storage trie accounts (with the given root) from db.
+// Intended for use when the trie with root has completed syncing.
+func ClearSyncStorageTrie(db ethdb.KeyValueStore, root common.Hash) error {
+	accountsPrefix := make([]byte, len(syncStorageTriesPrefix)+common.HashLength)
+	copy(accountsPrefix, syncStorageTriesPrefix)
+	copy(accountsPrefix[len(syncStorageTriesPrefix):], root[:])
+	return clearPrefix(db, accountsPrefix, syncStorageTriesKeyLength)
+}
+
+// ClearAllSyncStorageTries removes all storage tries added for syncing from db
+func ClearAllSyncStorageTries(db ethdb.KeyValueStore) error {
+	return clearPrefix(db, syncStorageTriesPrefix, syncStorageTriesKeyLength)
+}
+
+// ParseSyncStorageTrieKey returns the root and account for a storage trie
+// key returned from NewSyncStorageTriesIterator.
+func ParseSyncStorageTrieKey(keyBytes []byte) (common.Hash, common.Hash) {
+	keyBytes = keyBytes[len(syncStorageTriesPrefix):] // skip prefix
+	root := common.BytesToHash(keyBytes[:common.HashLength])
+	account := common.BytesToHash(keyBytes[common.HashLength:])
+	return root, account
+}
+
+// WriteSyncPerformed logs an entry in `db` indicating the VM state synced to `blockNumber`.
+func WriteSyncPerformed(db ethdb.KeyValueWriter, blockNumber uint64) error {
+	syncPerformedPrefixLen := len(syncPerformedPrefix)
+	bytes := make([]byte, syncPerformedKeyLength)
+	copy(bytes[:syncPerformedPrefixLen], syncPerformedPrefix)
+	binary.BigEndian.PutUint64(bytes[syncPerformedPrefixLen:], blockNumber)
+	return db.Put(bytes, []byte{0x01})
+}
+
+// NewSyncPerformedIterator returns an iterator over all block numbers the VM
+// has state synced to.
+func NewSyncPerformedIterator(db ethdb.Iteratee) ethdb.Iterator {
+	return rawdb.NewKeyLengthIterator(db.NewIterator(syncPerformedPrefix, nil), syncPerformedKeyLength)
+}
+
+// ParseSyncPerformedKey returns the block number from keys the iterator returned
+// from NewSyncPerformedIterator.
+func ParseSyncPerformedKey(key []byte) uint64 {
+	return binary.BigEndian.Uint64(key[len(syncPerformedPrefix):])
+}
+
+// GetLatestSyncPerformed returns the latest block number state synced performed to.
+func GetLatestSyncPerformed(db ethdb.Iteratee) (uint64, error) {
+	it := NewSyncPerformedIterator(db)
+	defer it.Release()
+
+	var latestSyncPerformed uint64
+	for it.Next() {
+		syncPerformed := ParseSyncPerformedKey(it.Key())
+		if syncPerformed > latestSyncPerformed {
+			latestSyncPerformed = syncPerformed
+		}
+	}
+	return latestSyncPerformed, it.Error()
+}
+
+// clearPrefix removes all keys in db that begin with prefix and match an
+// expected key length. `keyLen` must include the length of the prefix.
+func clearPrefix(db ethdb.KeyValueStore, prefix []byte, keyLen int) error {
+	it := db.NewIterator(prefix, nil)
+	defer it.Release()
+
+	batch := db.NewBatch()
+	for it.Next() {
+		key := common.CopyBytes(it.Key())
+		if len(key) != keyLen {
+			continue
+		}
```

Comment:
No point copying `key`s if we aren't going to hold a reference to them during the next iteration.
```suggestion
		key := it.Key()
		if len(key) != keyLen {
			continue
		}
		key = common.CopyBytes(key)
```